include/math/mls_polyinterpolation.tcc

```diff
@@ -0,0 +1,148 @@
+    else if (distance(dim) > 0.5 && distance(dim) < 1.0)
+      weights(dim) = 4. / 3. - (4. * distance(dim)) +
+                     (4. * distance(dim) * distance(dim)) -
+                     (4. * distance(dim) * distance(dim) * distance(dim) / 3.);
+    else
+      weights(dim) = 0.;
+  }
+  return weights.prod();
+}
+
+// Return quadratic spline weight of the data point
+template <unsigned Tdim, unsigned Tnmonomials>
+double mpm::MLSPolyInterpolation<Tdim, Tnmonomials>::quadratic_spline_weight(
+    const Eigen::Matrix<double, Tdim, 1>& distance) const {
+
+  Eigen::Matrix<double, Tdim, 1> weights;
+  weights.setZero();
+  for (unsigned dim = 0; dim < Tdim; ++dim) {
+    if (distance(dim) <= 0.5)
+      weights(dim) = 3. / 4. - (distance(dim) * distance(dim));
+    else if (distance(dim) > 0.5 && distance(dim) < 1.5)
+      weights(dim) = 0.5 * (1.5 - distance(dim)) * (1.5 - distance(dim));
+    else
+      weights(dim) = 0.;
+  }
+  return weights.prod();
+}
+
+// Initialise M matrix
+template <unsigned Tdim, unsigned Tnmonomials>
+void mpm::MLSPolyInterpolation<Tdim, Tnmonomials>::initialise_M_matrix(
+    const std::vector<VectorDim>& data_points, unsigned poly_order) {
+
+  M_matrix_.setZero();
+  unsigned p = 0;
+  for (const auto& pdata_coord : data_points) {
+    // Monomials
+    Eigen::VectorXd monomials =
+        mpm::Polynomial::evaluate_monomials<Tdim>(poly_order, pdata_coord);
+    // Compute M matrix
+    this->M_matrix_ += this->weights_[p] * monomials * monomials.transpose();
+    ++p;
+  }
+  if (!std::fabs(M_matrix_.determinant()) > 0.)
+    throw std::runtime_error("MLS M-matrix is non-invertible!");
+}
+
+// Initialise B vector
+template <unsigned Tdim, unsigned Tnmonomials>
+void mpm::MLSPolyInterpolation<Tdim, Tnmonomials>::initialise_B_vector(
+    const std::vector<VectorDim>& data_points, unsigned poly_order) {
+
+  B_vector_.clear();
+  unsigned p = 0;
+  for (const auto& pdata_coord : data_points) {
+    // Monomials
+    Eigen::VectorXd monomials =
+        mpm::Polynomial::evaluate_monomials<Tdim>(poly_order, pdata_coord);
+    // Compute M matrix
+    B_vector_.emplace_back(weights_[p] * monomials);
+    ++p;
+  }
+}
+
+// Compute polynomial interpolation at the point of interest
+template <unsigned Tdim, unsigned Tnmonomials>
+double mpm::MLSPolyInterpolation<Tdim, Tnmonomials>::interpolate_polynomial(
+    const std::vector<double>& data_values) const {
+  double pvalue = std::numeric_limits<double>::max();
+
+  if (data_values.size() != weights_.size() ||
+      data_values.size() != B_vector_.size())
+    throw std::runtime_error("MLS Data points are invalid");
+
+  // B Matrix
+  Eigen::Matrix<double, Tnmonomials, 1> B_matrix;
+  B_matrix.setZero();
+  for (unsigned i = 0; i < data_values.size(); ++i)
+    B_matrix += data_values[i] * B_vector_[i];
+
+  // Compute interpolated value at the point of interest
+  pvalue = point_monomials_.transpose() * M_matrix_.inverse() * B_matrix;
```

Comment:
M_matrix_ is inversed multiple time as you are calling interpolate_polynomial, wouldn't it be better to store M_matrix_.inverse() as I see that M_matrix_ is constant as it is initialized.